### ISSUE TITLE
updated List equality produces false negative

### ISSUE THREAD

**Issue by dustin-graham** [felangel/equatable#19]
**Describe the bug**
The new list equality check introduces false negatives in my tests when comparing identical lists with different runtime types. In consequence of other business logic in my code, I end up with a `GrowableList` in one side of the equality and a `CopyOnWriteList` on the other. Contents are identical but the runtime check fails.

**To Reproduce**
In my test I generate dummy data with `List.generate()`. This produces a `GrowableList`. However in my code I am using `BuildList` which has a `toList()` method that returns a `CopyOnWriteList`. The equality check worked as expected with older code.

**Expected behavior**
Deep equals check should find that the list contents are identical

**Screenshots**
<img width="1015" alt="Screen Shot 2019-05-27 at 7 46 53 AM" src="https://user-images.githubusercontent.com/959931/58424002-b04d2f00-8053-11e9-8bbe-22883694e28c.png">

**Version**
Dart VM version: 2.3.0 (Fri May 3 10:32:31 2019 +0200) on "macos_x64"

**Additional context**
I'm able to work around this in my tests by wrapping my original dummy data in a `BuiltList` and using the `toList()` to compare. But this is a bit more work than I would expect and might be confusing to future maintainers. Thanks!

**Comment by felangel** [comment]
Hi @dustin-graham 👋 
Thanks for opening an issue!

I'll try to address the issue in the next few days. Alternatively, if you want/have the time you are more than welcome to open a PR 👍

**Comment by dustin-graham** [comment]
@felangel thanks for the reply. I'd be happy to contribute. I'd like to be sure I understand the motivations for this change so that I don't cause someone else an unexpected change. Could we safely just remove the runtimeType check?

**Comment by jorgecoca** [comment]
I think the library is working as expected. It is our requirement that, in order to assert that two objects are equal, their runtimes must be equal as well. 
Otherwise, we might end up in situations where `Person(name: 'Jorge')` and `Duck(name: 'Jorge')` might be the same object.
If runtime is not necessary for your equality requirements, you can always implement manually `==` and `hashCode` to have more control over the conditions.

**Comment by dustin-graham** [comment]
@jorgecoca I agree for the most part. However, in this spot, we're comparing two lists as given by the function signature. This runtime type check is asserting that the two lists are of the same type, not whether their contents are equal.

https://github.com/felangel/equatable/blob/f8f9b32259742be0f772bf958d919c354ea71792/lib/src/equatable_utils.dart#L38

**Comment by felangel** [comment]
@dustin-graham I agree with @jorgecoca. I would actually consider the fact that this worked in the previous release a bug. If the `runtimeTypes` are different then I don't think we can consider the two objects equal because this would introduce many other bugs like @jorgecoca mentioned earlier.

**Comment by dustin-graham** [comment]
@felangel @jorgecoca I put together a small sample to illustrate this issue

https://github.com/dustin-graham/dart_equality_sample

I'll inline the relevant code here:
```dart
main(List<String> arguments) {
  final growableList = List.generate(1, (_) {
    return Student((b) {
      b.name = "Student 1";
      b.email = "1@student.com";
    });
  });
  final builtStudentList = BuiltList<Student>(growableList).toList();

  final listEquality = ListEquality<Student>();
  final collectionEquality =
      listEquality.equals(growableList, builtStudentList);
  print("collection lib equality: $collectionEquality"); // prints true

  // pass in the lists as if given via a props list
  final equatableEquals =
      EquatableUtils.equals([growableList], [builtStudentList]);
  print("equatable lib equality: $equatableEquals"); // prints false

  print("built list is List: ${builtStudentList.toList() is List<Student>}");
  print("regular list is List: ${growableList is List<Student>}");
  var runtimeType = builtStudentList.runtimeType;
  print("built list runtime type: ${runtimeType}");
  var runtimeType2 = growableList.runtimeType;
  print("generated list runtime type: ${runtimeType2}");
}
```
I made a PR that fixes the issue and keeps all the runtime type checking. 
https://github.com/felangel/equatable/pull/20

The line that checks if the units are collections is the correct special-case runtime type check for collection types. The issue is that it wasn't being reached because of the general purpose runtimeType check right before it. The sample code demonstrates that two lists of different runtime types are considered equal in the Collection package, but not even compared in Equatable. This PR simply puts the collection runtime type check first and the general purpose runtime type check second. I believe this is the correct order. Thanks!

**Comment by algodave** [comment]
@felangel I think the changes added between `0.2.4` and `0.2.5` are relevant, I think they deserve crafting the `0.3.0` (btw: Releases here on GitHub are stuck on `0.2.3`)

**Comment by felangel** [comment]
@algodave what do you mean they deserve crafting `0.3.0`? We're following semantic versioning and this was a bug fix in my opinion without a breaking API change. Can you please elaborate?

Also, thanks for pointing out the lack of 0.2.4 and 0.2.5 releases! I have updated the releases now 👍

**Comment by algodave** [comment]
@felangel Deciding whether 2 lists having the same elements but a different `runtimeType` should be equal or not could be also considered a design decision rather than a bug 🙂 - up to you of course!

In my case this change was "breaking" because all my Bloc tests are failing with `0.2.5`. They use the following pattern:

```dart
test('My Bloc event test', () {
  // mock stuff etc.
  final expectedStates = [MyState(), MyOtherState()];
  expectLater(
    blocBeingTested.state, 
    emitsInOrder(expectedStates),
  );
  blocBeingTested.dispatch(MyEvent());
});
```

and the error I get is that the expected list of states is not equal to the one emitted by the bloc being tested.

**Comment by felangel** [comment]
@algodave I don't believe this change should have broken your tests. Are you able to share a sample that worked with equatable v0.2.4 and breaks with v0.2.5?

**Comment by algodave** [comment]
@felangel Thank you for answering, it helped me to look better at my tests, and the problem was there! 

Some of my states hold a `List<MyModel>` prop; in tests, I defined `MyModelMock extends Mock implements MyModel` using `mockito`. 
Lists in expected states were instantiated as `List<MyModelMock>` which caused the equality check fail. I changed them to `List<MyModel>` and everything works fine.

So my app is now able to run `v0.2.5` joyfully 🙂
